logback/src/main/java/com/linecorp/armeria/common/logback/RequestContextExportingAppender.java

```diff
@@ -0,0 +1,388 @@
+ * {@link RequestContext} to {@link MDC}. It can export any of the following properties.
+ *
+ * <h3>Built-in properties</h3>
+ * A built-in property is a protocol-agnostic property available for most requests. See the complete list of
+ * the built-in properties and their MDC keys at {@link BuiltInProperty}.
+ * Use {@link #addBuiltIn(BuiltInProperty)} to add them programmatically.
+ *
+ * <h3>HTTP request and response headers</h3>
+ * When the current {@link SessionProtocol} is HTTP, a user can export HTTP headers of the current request and
+ * response. Use {@link #addHttpRequestHeader(CharSequence)} and {@link #addHttpResponseHeader(CharSequence)}
+ * to add them programmatically. The MDC key of the exported header is
+ * {@code "req.http_headers.<lower-case header name>"} or {@code "res.http_headers.<lower-case header name>"}.
+ * e.g. {@code "req.http_headers.user-agent"} and {@code "res.http_headers.set-cookie"}
+ *
+ * <h3>Custom attributes</h3>
+ * A user can attach an arbitrary {@link Attribute} to a {@link RequestContext} and export its value.
+ * Use {@link #addAttribute(String, AttributeKey)} to add them programmatically. The MDC key of the exported
+ * attribute is {@code "attrs.<alias>"}.
+ *
+ * <h3>Configuring with XML or {@link JoranConfigurator}</h3>
+ * Specify the MDC key of the property with an {@code <export />} tag:
+ * <pre>{@code
+ * <?xml version="1.0" encoding="UTF-8"?>
+ * <configuration>
+ *   <appender name="SOME_APPENDER" class="...">
+ *     ...
+ *   </appender>
+ *   <appender name="RCIA" class="com.linecorp.armeria.common.logback.RequestContextExportingAppender">
+ *     <appender-ref ref="SOME_APPENDER" />
+ *     <export>remote.ip</export>
+ *     <export>remote.port</export>
+ *     <export>req.http_headers.user-agent</export>
+ *     <export>res.http_headers.set-cookie</export>
+ *     <export>attrs.req_id:com.example.RequestId#ATTR</export>
+ *   </appender>
+ *   ...
+ * </configuration>
+ * }</pre>
+ *
+ * <p>Note that custom attributes are slightly different from other types of properties. You need to
+ * concatenate {@linkplain AttributeKey#name() the full name of the attribute} to the MDC key, with a colon.
+ */
+public class RequestContextExportingAppender extends UnsynchronizedAppenderBase<ILoggingEvent>
+                                             implements AppenderAttachable<ILoggingEvent> {
+
+    private static final String PREFIX_ATTRS = "attrs.";
+    private static final String PREFIX_HTTP_REQ_HEADERS = "req.http_headers.";
+    private static final String PREFIX_HTTP_RES_HEADERS = "res.http_headers.";
+
+    private static final String STAMP = "__armeria_stamp";
+
+    private RequestContextExporter exporter;
+
+    private final AppenderAttachableImpl<ILoggingEvent> aai = new AppenderAttachableImpl<>();
+    private final Set<BuiltInProperty> builtIns = EnumSet.noneOf(BuiltInProperty.class);
```

Comment:
We cannot fix the mutability of appender because XML configuration requires it to be so - it has to call `setExport()` on each `<export/>` element. Please let me know if there's a way to build an immutable appender using Joran.